File: mac/Classes/Model/Actions/Specific/UserScriptAction.m

#import "UserScriptAction.h"
#import "Project.h"
#import "LRProjectTargetResult.h"


@implementation UserScriptAction

- (NSString *)label {
    return [NSString stringWithFormat:NSLocalizedString(@"Run %@", nil), self.scriptName];
}

+ (NSSet *)keyPathsForValuesAffectingLabel {
    return [NSSet setWithObject:@"scriptName"];
}

- (void)invokeForProject:(Project *)project withModifiedFiles:(NSArray *)files result:(LROperationResult *)result completionHandler:(dispatch_block_t)completionHandler {
    [self.script invokeForProjectAtPath:project.rootURL.path withModifiedFiles:[NSSet setWithArray:[files valueForKeyPath:@"relativePath"]] result:result completionHandler:completionHandler];
}

- (UserScript *)script {
    if (_scriptName.length == 0)
        return nil;

    NSArray *userScripts = [UserScriptManager sharedUserScriptManager].userScripts;
    for (UserScript *userScript in userScripts) {
        if ([userScript.uniqueName isEqualToString:_scriptName])
            return userScript;
    }

    return [[MissingUserScript alloc] initWithName:_scriptName];
}

+ (NSSet *)keyPathsForValuesAffectingScript {
    return [NSSet setWithObject:@"scriptName"];
}

- (BOOL)isNonEmpty {
    return self.scriptName.length > 0 && self.script.exists;
}

+ (NSSet *)keyPathsForValuesAffectingNonEmpty {
    return [NSSet setWithObject:@"scriptName"];
}

- (void)loadFromMemento:(NSDictionary *)memento {
    [super loadFromMemento:memento];
    self.scriptName = memento[@"script"] ?: @"";
}

- (void)updateMemento:(NSMutableDictionary *)memento {
    [super updateMemento:memento];
    memento[@"script"] = self.scriptName;
}

- (LRTarget *)targetForModifiedFiles:(NSArray *)files {
    if ([self inputPathSpecMatchesFiles:files]) {
        return [[LRProjectTargetResult alloc] initWithAction:self modifiedFiles:files];
    } else {
        return nil;
    }
}

@end

File: mac/Classes/Model/Actions/Specific/UserScriptAction.swift

import Foundation

@objc(UserScriptAction)
class UserScriptAction : Action {

    var scriptName: String?

    var script: UserScript? {
        if let actualScriptName = scriptName {
            let userScripts = UserScriptManager.sharedUserScriptManager().userScripts as UserScript[]
            if let matchingScript = findIf(userScripts, { $0.uniqueName == actualScriptName }) {
                return matchingScript
            } else {
                return MissingUserScript(name: actualScriptName)
            }
        } else {
            return nil;
        }
    }

    override var label: String {
        return "Run \(scriptName)"
    }

    class func keyPathsForValuesAffectingLabel() -> NSSet {
        return NSSet(object: "scriptName")
    }

    class func keyPathsForValuesAffectingScript() -> NSSet {
        return NSSet(object: "scriptName")
    }

    class func keyPathsForValuesAffectingNonEmpty() -> NSSet {
        return NSSet(object: "scriptName")
    }

    override var nonEmpty: Bool {
        if !scriptName {
            return false
        } else if let actualScript = script {
            return actualScript.exists
        } else {
            return false
        }
    }

    override func loadFromMemento() {
        super.loadFromMemento()
//        scriptName = EmptyToNil(memento["script"] as? String)
    }

    override func updateMemento() {
        super.updateMemento()
        memento["script"] = scriptName
    }

    override func targetForModifiedFiles(files: LRProjectFile[]) -> LRTarget? {
        if inputPathSpecMatchesFiles(files) {
            return LRProjectTarget(action: self, modifiedFiles: files as LRProjectFile[])
        } else {
            return nil
        }
    }

    override func invokeWithModifiedFiles(files: LRProjectFile[], result: LROperationResult, completionHandler: dispatch_block_t) {
        let trueFiles = files as LRProjectFile[]
        let filePaths = trueFiles.map { $0.relativePath }

        if let actualScript = script {
            actualScript.invokeForProjectAtPath(project.rootURL.path, withModifiedFiles:NSSet(array: filePaths), result: result, completionHandler: completionHandler)
        }
    }

}
